src/GafferUI/GraphGadget.cpp

```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
+		return;
+	}
+
+
+	tbb::spin_mutex &pendingActiveStateMutex = m_pendingActiveStateMutex;
+	std::unique_ptr<PendingActive> &pendingActiveState = m_pendingActiveState;
+
+	// Must hold a reference to stop us dying before our UI thread call is scheduled.
+	ConstGraphGadgetPtr thisRef = this;
+
+	m_activeStateTask = Gaffer::ParallelAlgo::callOnBackgroundThread(
+		// Subject:  TODO - does it matter that we're only including one plug as a subject?  We respond to any
+		// plugs being dirtied on the focus node with manual cancellation of this task, so hopefully it should
+		// be OK
+		focusPlugs[0].get(),
+		[focusPlugs, &pendingActiveStateMutex, &pendingActiveState, thisRef ] {
+			// We don't use the current context for our actual context - we use the script context
+			// and activePlugsAndNodes needs to continue even in cases where it can't figure out an accurate
+			// context andt just uses null ... but we do need to take the canceller from the context
+			// provided by callOnBackgroundThread
+			const IECore::Canceller *canceller = Gaffer::Context::current()->canceller();
+
+			Canceller::check( canceller );
+
+			std::unique_ptr<PendingActive> newActive( new PendingActive );
+			const Gaffer::ScriptNode *script = focusPlugs[0]->ancestor<Gaffer::ScriptNode>();
+			const Gaffer::Context *context = script ? script->context() : nullptr;
+
+			for( const Gaffer::ConstPlugPtr &focusPlug : focusPlugs )
+			{
+				activePlugsAndNodes(
+					focusPlug.get(), context, canceller,
+					newActive->activePlugs, newActive->activeNodes
+				);
+			}
+
+			{
+				tbb::spin_mutex::scoped_lock scopedLock( pendingActiveStateMutex );
+				pendingActiveState.swap( newActive );
+			}
+
+			Gaffer::ParallelAlgo::callOnUIThread(
+				[thisRef] {
+					thisRef->applyActive( true );
+				}
+			);
+		}
+	);
+
+}
+
+void GraphGadget::applyActive( bool hasFocus ) const
+{
+	std::unique_ptr<PendingActive> pendingActiveState;
+	{
+		tbb::spin_mutex::scoped_lock scopedLock( m_pendingActiveStateMutex );
+		m_pendingActiveState.swap( pendingActiveState );
+	}
+
+
+	if( hasFocus )
+	{
+		if( pendingActiveState )
+		{
+			for( auto &g : children() )
+			{
+				if( ConnectionGadget *c = runTimeCast<ConnectionGadget>( g.get() ) )
+				{
+					c->activeForFocusNode( pendingActiveState->activePlugs.count( c->dstNodule()->plug() ) );
+				}
+				else if( NodeGadget *n = runTimeCast<NodeGadget>( g.get() ) )
+				{
+					n->activeForFocusNode( pendingActiveState->activeNodes.count( n->node() ) );
+				}
+			}
+		}
+	}
+	else
+	{
+		// If we haven't got a focus node to trigger active state visualisation, just show everything active
+		for( auto &g : children() )
+		{
+			if( ConnectionGadget *c = runTimeCast<ConnectionGadget>( g.get() ) )
+			{
+				c->activeForFocusNode( true );
+			}
+			else if( NodeGadget *n = runTimeCast<NodeGadget>( g.get() ) )
+			{
+				n->activeForFocusNode( true );
+			}
+		}
+	}
+
+	m_activeStateDirty = false;
+
+	// TODO - this const_cast is pretty ugly
```

Comment:
Seems like `updateActive()` and `applyActive()` should be non-const since they do mutate stuff, and the source of the problem is `updateActive()` being called from `doRenderLayer()`? Maybe the ugly cast belongs there, perhaps with a comment about us being in trouble if ViewportGadget changes drawing strategy to include display lists? I dunno, given that we're thinking we might overhaul the activation mechanism shortly anyway, maybe it's not worth worrying.

Reply:
Addressed in 2ddd2b2e

I'm not sure a display list drawing strategy specifically is the problem - we would still need to rebuild the display lists when the graph is edited, and rebuilding the display list could trigger the activation walk.  It's maybe a little odd to wait for the display list rebuild before starting to think about calculating active, and needing another display list rebuild when it finishes, but the same is basically true of the current code if you're not panning around to require rerendering.